iceoryx_posh/source/mepoo/chunk_header.cpp

```diff
@@ -21,8 +22,74 @@ namespace iox
 {
 namespace mepoo
 {
-ChunkHeader::ChunkHeader() noexcept
+ChunkHeader::ChunkHeader(const uint32_t chunkSize,
+                         const uint32_t payloadSize,
+                         const uint32_t payloadAlignment,
+                         const uint32_t customHeaderSize,
+                         const uint32_t customHeaderAlignment) noexcept
 {
+    static_assert(alignof(ChunkHeader) >= 8U,
+                  "All the calculations expect the ChunkHeader alignment to be at least 8!");
+    cxx::Expects(customHeaderAlignment <= alignof(ChunkHeader)
+                 && "The alignment of the custom header must not exceed the alignment of the ChunkHeader!");
+
+    this->chunkSize = chunkSize;
+    this->payloadSize = payloadSize;
+
+    // have a look at »Payload Offset Calculation« in chunk_header.md for more details regarding the calculation
+    if (customHeaderSize == 0U)
+    {
+        if (payloadAlignment <= alignof(mepoo::ChunkHeader))
+        {
+            // the most simple case with no custom header and the payload adjacent to the ChunkHeader
+            payloadOffset = sizeof(ChunkHeader);
+        }
+        else
+        {
+            // the second most simple case with no custom header but the payload alignment
+            // exceeds the ChunkHeader alignment and is therefore not necessarily adjacent
+            uint64_t addressOfChunkHeader = reinterpret_cast<uint64_t>(this);
+            uint64_t headerEndAddress = addressOfChunkHeader + sizeof(ChunkHeader);
+            uint64_t alignedPayloadAddress = iox::cxx::align(headerEndAddress, static_cast<uint64_t>(payloadAlignment));
+            uint64_t offsetToPayload = alignedPayloadAddress - addressOfChunkHeader;
+            cxx::Ensures(offsetToPayload <= std::numeric_limits<PayloadOffset_t>::max()
+                         && "Payload offset too large for chunk!");
+            payloadOffset = static_cast<PayloadOffset_t>(offsetToPayload);
+
+            // this is safe since the alignment of the payload is larger than the one from the ChunkHeader
+            // -> the payload is either adjacent and `backOffset` is at the same location as `payloadOffset`
+            //    or the payload is not adjacent and there is space of at least the alignment of ChunkHeader
+            //    between the ChunkHeader and the payload
+            auto addressOfBackOffset = alignedPayloadAddress - sizeof(PayloadOffset_t);
+            auto backOffset = reinterpret_cast<PayloadOffset_t*>(addressOfBackOffset);
+            *backOffset = payloadOffset;
+        }
+    }
+    else
+    {
+        // the most complex case with a custom header
+        auto addressOfChunkHeader = reinterpret_cast<uint64_t>(this);
+        uint64_t headerEndAddress = addressOfChunkHeader + sizeof(ChunkHeader) + customHeaderSize;
+        uint64_t potentialBackOffsetAddress =
```

Comment:
I think the name `unalignedBackOffsetAddress` is better fitting.

Reply:
well, it is perfectly aligned. The problem is that it must be adjacent to the payload and the payload might have a different alignment. The actual address can only be calculated later on, so this is only the potential address. Only after we know the aligned address of the payload we can recalculate the back offset address.

Reply:
what do you think about `anticipatedBackOffsetAddress`?